CMakeLists.txt

```diff
@@ -0,0 +1,150 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2013-2022 The Foundry Visionmongers Ltd
+
+#-----------------------------------------------------------------------
+# CMake preamble.
+
+# FindPython's Development.Module component was added in 3.18, and
+# pybind11 recommends 3.18.2+.
+cmake_minimum_required(VERSION 3.18.2)
+
+# Additional include directories for CMake utils.
+set(CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake)
+
+if (APPLE AND NOT DEFINED CMAKE_OSX_DEPLOYMENT_TARGET)
+    # The value of this variable should be set prior to the first
+    # project() command invocation because it may influence
+    # configuration of the toolchain and flags.
+    set(CMAKE_OSX_DEPLOYMENT_TARGET "10.13" CACHE STRING "Minimum OS X deployment version")
+endif ()
+
+# Forbid in-source builds.
+set(CMAKE_DISABLE_IN_SOURCE_BUILD ON)  # NB: currently undocumented.
+# Forbid CMake from editing sources
+set(CMAKE_DISABLE_SOURCE_CHANGES ON)  # NB: currently undocumented.
+
+# By default, do not output "Up-to-date" messages during install. Can
+# be spammy when this project is included in another via CMake's
+# ExternalProject.
+if (NOT DEFINED CMAKE_INSTALL_MESSAGE)
+    set(CMAKE_INSTALL_MESSAGE LAZY)
+endif ()
+
+
+#-----------------------------------------------------------------------
+# Project metadata
+
+string(CONCAT project_description
+    "An open-source interoperability standard for tools and content"
+    " management systems used in media production.")
+
+project(
+    OpenAssetIO
+    VERSION 0.0.0
+    DESCRIPTION ${project_description}
+    HOMEPAGE_URL https://thefoundryvisionmongers.github.io/OpenAssetIO
+    LANGUAGES C CXX)
+
+
+#-----------------------------------------------------------------------
+# Global settings
+
+include(ProjectSettings)
+
+
+#-----------------------------------------------------------------------
+# Options
+
+# Installation options
+# TODO(DF): Install dir wrangling depending on desired Python/C++
+#   library combination. See OTIO.
+
+# Build options
+# TODO(DF): This means CMake controls whether static or shared libs are
+#   created with `add_library` when no lib type is specified. Perhaps
+#   we shouldn't use this built-in CMake variable, since it can affect
+#   dependencies too (e.g.
+#   https://github.com/boostorg/boost_install/issues/5). See also OTIO.
+option(BUILD_SHARED_LIBS "Set to OFF to build static libraries" ON)
```

Comment:
I had assumed that in CMake GUIs the `BUILD_SHARED_LIBS` would be hidden by default, and adding an `option` would mark it as non-advanced?  This option was lifted from one of the other open source projects I'd used as a reference, so I'm not sure about it, hence the comment. 

> It actually only matters if you do add_library and not specify the library type (meaning it can be built as either shared or static).
Yeah as mentioned in the comment.  I'm not sure if that's a good pattern or if we should be explicit.

Reply:
I see it passed on Conan generated CMake command lines, for a large variety of thirdparty projects, and I don't recall seeing it written down as a option in the scripts. I might've completely overlooked it, but I'm pretty sure an option isn't needed.

Regarding explicit, it depends.
If you want to generate one type of library and that's it, then I prefer explicit.

If you want to be flexible and build either type of library (static or shared), then if you are explicit in the target specification, you need two targets. This variable is an optimisation to have just one target in your scripts, and CMake flips the type under the hood.

Reply:
Ultimately, I think we need to allow the users of the project to easily decide if it is building a static or shared lib as they are the ones who know how it is going to be used. We may have an opinion about our specific builds for CI/distribution, but someone with the source tree should be able to do what they need with minimal fuss.

Reply:
If you're offering a pure C++ library option alone (i.e. no Python bindings), I agree to give consumers the choice.

If you've always got Python in play, I think you may be forced down the shared library route.

Reply:
It's layers though. The python bindings are an optional build artefact. If someone wants to make a static/shared OpenAssetIO Core C++ lib on its own, in isolation, they should be able to. If they want to also build the C++ plugin system, they should be able to do so. etc...

It may well be that most common workflows require all four things (core, plugin system, core python, python plugin system) - as the target asset manager is only written in python (our own CI will be no exception), but they should all be individual targets that can be build as needed.

Reply:
The [cmake docs](https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html) imply its useful to add as an `option`

>  This variable is often added to projects as an option() so that each user of a project can decide if they want to build the project using shared or static libraries.

As I suspected, I just confirmed that the effect of this is to "surface" the option in GUIs (I tried `ccmake`)  - without adding the `option` the `BUILD_SHARED_LIBS` variable is hidden by default.

But perhaps, based on the discussion above, we need to also allow building _both_ static and shared at the same time, in which case we'll need to add a custom `BUILD_BOTH` option, and be much more explicit when creating the (now two) target(s).  We can defer this for now, though(?)

Reply:
Do what must be done (said the Emperor).

I did just want to check that what I originally said wasn't nonsense, so have a look at https://gist.github.com/markfinal/18870327dc611a34a937098ccbb76541 for details

Reply:
After conversation in the working group, I removed the TODO in a2d9b7e9.  Then I added a comment saying why it's declared as an `option`, as well as a TODO re. simultaneously building shared and static libs, in 1c32585a